### PR TITLE
Ewso365 fix fetch incidents

### PR DESCRIPTION
**Pull request by JasBeilin**
## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://github.com/demisto/etc/issues/38869

## Description
Fixed an issue where fetching incidents with the same time (while max_fetch is lower than the number of incidents) got stuck in infinite loop, since the last run was taken from last incident, regardless whether it was excluded (already processed) or not.

## Does it break backward compatibility?
   - [x] No

## Must have
- [x] Tests
- [x] Documentation

### CODE REVIEW

**Comment by JasBeilin** [comment]
When running on master (bug exist):
image

The First fetch got the first result, the second one got 0 incidents as the one it found already was excluded.

File: Packs/EWS/Integrations/EWSO365/EWSO365.py [hunk 1]
```diff
@@ -2143,33 +2143,38 @@ def fetch_emails_as_incidents(client: EWSClient, last_run):
...
 
+        demisto.debug(f'{APP_NAME} - ending fetch - got {len(incidents)} incidents.')
         last_run_time = incident.get("occurred", last_run.get(LAST_RUN_TIME))
         if isinstance(last_run_time, EWSDateTime):
             last_run_time = last_run_time.ewsformat()
 
+        if last_run_time > LAST_RUN_TIME:
+            ids = current_fetch_ids
+        else:
+            ids = current_fetch_ids | excluded_ids
```
Comment (anara123) [inline Packs/EWS/Integrations/EWSO365/EWSO365.py:2176]: add a comment that explains the logic here.It is not clear to me what this code does and why

File: Packs/EWS/Integrations/EWSv2/EWSv2.py [hunk 2]
```diff
@@ -917,23 +917,21 @@ def fetch_last_emails(account, folder_name='Inbox', since_datetime=None, exclude
     qs = qs.filter().order_by('datetime_received')
 
     result = []
+    counter = 0
     for item in qs:
+        counter += 1
         try:
-            if isinstance(item, Message):
+            if isinstance(item, Message) and item.message_id not in exclude_ids:
```
Comment (esharf) [inline Packs/EWS/Integrations/EWSv2/EWSv2.py:924]:
```suggestion
            if isinstance(item, Message) and not item.message_id in exclude_ids:
```
Better performance.

File: Packs/EWS/Integrations/EWSO365/EWSO365.py [hunk 3]
```diff
@@ -2143,33 +2143,38 @@ def fetch_emails_as_incidents(client: EWSClient, last_run):
...
 
                 if len(incidents) >= client.max_fetch:
                     break
 
+        demisto.debug(f'{APP_NAME} - ending fetch - got {len(incidents)} incidents.')
         last_run_time = incident.get("occurred", last_run.get(LAST_RUN_TIME))
         if isinstance(last_run_time, EWSDateTime):
             last_run_time = last_run_time.ewsformat()
 
+        if last_run_time > LAST_RUN_TIME:
```
Comment (anara123) [inline Packs/EWS/Integrations/EWSO365/EWSO365.py:2173]: You compare last_run_time which is time with `LAST_RUN_TIME` which is `"lastRunTime"`.

File: Packs/EWS/Integrations/EWSv2/EWSv2.py [hunk 4]
```diff
@@ -1241,34 +1239,42 @@ def parse_incident_from_item(item, is_fetch):
...
-            LAST_RUN_TIME: last_run_time,
-            LAST_RUN_FOLDER: folder_name,
-            LAST_RUN_IDS: list(ids),
-            ERROR_COUNTER: 0
-        }
+                LAST_RUN_TIME: last_run_time,
+                LAST_RUN_FOLDER: folder_name,
+                LAST_RUN_IDS: list(ids),
+                ERROR_COUNTER: 0
+            }
```
Comment (esharf) [inline Packs/EWS/Integrations/EWSv2/EWSv2.py:1277]: remove the indentation.

File: Packs/EWS/Integrations/EWSO365/EWSO365.py [hunk 5]
```diff
@@ -2215,15 +2220,17 @@ def fetch_last_emails(
...
+        counter += 1
+        if isinstance(item, Message) and item.message_id not in exclude_ids:
             result.append(item)
             if len(result) >= client.max_fetch:
                 break
-
-    if exclude_ids and len(exclude_ids) > 0:
-        exclude_ids = set(exclude_ids)
-        result = [x for x in result if x.message_id not in exclude_ids]
+    demisto.debug(f'{APP_NAME} - Got total of {counter} from ews query. '
```
Comment (anara123) [inline Packs/EWS/Integrations/EWSO365/EWSO365.py:2232]: Check if you can print also the total number of emails and not only the ones we fetched

File: Packs/EWS/Integrations/EWSv2/EWSv2.py [hunk 6]
```diff
@@ -1241,34 +1239,42 @@ def parse_incident_from_item(item, is_fetch):
 
 def fetch_emails_as_incidents(account_email, folder_name):
     last_run = get_last_run()
+    excluded_ids = set(last_run.get(LAST_RUN_IDS))
```
Comment (esharf) [inline Packs/EWS/Integrations/EWSv2/EWSv2.py:1242]:
```suggestion
    excluded_ids = set(last_run.get(LAST_RUN_IDS, []))
```

File: Packs/EWS/Integrations/EWSO365/EWSO365.py [hunk 7]
```diff
@@ -2143,33 +2143,38 @@ def fetch_emails_as_incidents(client: EWSClient, last_run):
     :return:
     """
     last_run = get_last_run(client, last_run)
-
+    excluded_ids = set(last_run.get(LAST_RUN_IDS))
```
Comment (esharf) [inline Packs/EWS/Integrations/EWSO365/EWSO365.py:2146]:
```suggestion
    excluded_ids = set(last_run.get(LAST_RUN_IDS, []))
```

File: Packs/EWS/Integrations/EWSO365/EWSO365.py [hunk 8]
```diff
@@ -2215,15 +2220,17 @@ def fetch_last_emails(
     qs = qs.filter().order_by("datetime_received")
 
     result = []
+    exclude_ids = exclude_ids if exclude_ids else set()
```
Comment (esharf) [inline Packs/EWS/Integrations/EWSO365/EWSO365.py:2223]:
```suggestion
```
already handled in `fetch_emails_as_incidents` function

**Review by esharf — APPROVED** [review]
great work on UT

File: Packs/EWS/Integrations/EWSv2/EWSv2.py [hunk 9]
```diff
@@ -917,23 +917,21 @@ def fetch_last_emails(account, folder_name='Inbox', since_datetime=None, exclude
...
                     break
         except ValueError as exc:
             future_utils.raise_from(ValueError(
                 'Got an error when pulling incidents. You might be using the wrong exchange version.'
             ), exc)
-
-    if exclude_ids and len(exclude_ids) > 0:
-        exclude_ids = set(exclude_ids)
-        result = [x for x in result if x.message_id not in exclude_ids]
+    demisto.debug(f'EWS V2 - Got total of {counter} from ews query. {len(result)} results not excluded. ')
```
Comment (esharf) [inline Packs/EWS/Integrations/EWSv2/EWSv2.py:932]:
```suggestion
    demisto.debug('EWS V2 - Got total of {} from ews query. {} results not excluded. '.,fomat(counter, len(result)))
```
EWSv2 is python2 code

**Review by esharf — CHANGES_REQUESTED** [review]

**Review by esharf — APPROVED** [review]

File: Packs/EWS/Integrations/EWSv2/EWSv2.py (hunk 2 above)
Comment (JasBeilin) [reply Packs/EWS/Integrations/EWSv2/EWSv2.py:924]: https://www.flake8rules.com/rules/E713.html says not to.

Comment (esharf) [reply Packs/EWS/Integrations/EWSv2/EWSv2.py:924]: correct

File: Packs/EWS/Integrations/EWSO365/EWSO365.py (hunk 8 above)
Comment (JasBeilin) [reply Packs/EWS/Integrations/EWSO365/EWSO365.py:2223]: But fetch_last_email does not have to get it. Think it is better to keep it for testing of the specific function.

Comment (esharf) [reply Packs/EWS/Integrations/EWSO365/EWSO365.py:2223]: then add it also to EWSv2

File: Packs/EWS/ReleaseNotes/1_8_28.md [hunk 10]
```diff
@@ -0,0 +1,6 @@
+
+#### Integrations
+##### EWS O365
+- Fixed an issue when fetching incidents with lower max_fetch value then number of incidents with same creation time. 
+##### EWS v2
+- Fixed an issue when fetching incidents with lower max_fetch value then number of incidents with same creation time.
```
Comment (bakatzir) [inline Packs/EWS/ReleaseNotes/1_8_28.md:6]:
```suggestion
- Fixed an issue where fetching incidents with the same creation time missed to retrieve some of them. 
##### EWS v2
- Fixed an issue where fetching incidents with the same creation time missed to retrieve some of them. 
```

File: Packs/EWS/Integrations/EWSO365/EWSO365.py (hunk 8 above)
Comment (JasBeilin) [reply Packs/EWS/Integrations/EWSO365/EWSO365.py:2224]: I did :)

**Review by guykeller — APPROVED** [review]

File: Packs/EWS/Integrations/EWSO365/EWSO365.py [hunk 11]
```diff
@@ -2143,35 +2142,51 @@ def fetch_emails_as_incidents(client: EWSClient, last_run):
...
                     break
 
-        last_run_time = incident.get("occurred", last_run.get(LAST_RUN_TIME))
-        if isinstance(last_run_time, EWSDateTime):
-            last_run_time = last_run_time.ewsformat()
+        demisto.debug(f'{APP_NAME} - ending fetch - got {len(incidents)} incidents.')
+
+        last_fetch_time = last_run.get(LAST_RUN_TIME)
+
+        last_incident_run_time = incident.get("occurred", last_fetch_time)
```
Comment (yuvalbenshalom) [inline Packs/EWS/Integrations/EWSO365/EWSO365.py:2171]: @JasBeilin / @esharf  are we certain that items in the loop in 2158 are ordered by time, so the last incident in the loop is really the latest mail that was iterated by the loop?

File: Packs/EWS/Integrations/EWSv2/EWSv2.py [hunk 12]
```diff
@@ -1241,33 +1239,51 @@ def parse_incident_from_item(item, is_fetch):
...
                 if len(incidents) >= MAX_FETCH:
                     break
 
-        last_run_time = incident.get('occurred', last_run.get(LAST_RUN_TIME))
-        if isinstance(last_run_time, EWSDateTime):
-            last_run_time = last_run_time.ewsformat()
+        demisto.debug('EWS V2 - ending fetch - got {} incidents.'.format(len(incidents)))
+        last_fetch_time = last_run.get(LAST_RUN_TIME)
+
+        last_incident_run_time = incident.get("occurred", last_fetch_time)
```
Comment (yuvalbenshalom) [inline Packs/EWS/Integrations/EWSv2/EWSv2.py:1265]: same question about its, ordering as in O365

**Review by yuvalbenshalom — CHANGES_REQUESTED** [review]
see my comments in the code.
also, what unit test cover the code changes in EWSv2?
if you include the fix to 2 integrations, we should have tests in both.

File: Packs/EWS/Integrations/EWSv2/EWSv2.py (hunk 12 above)
Comment (JasBeilin) [reply Packs/EWS/Integrations/EWSv2/EWSv2.py:1263]: yes
image

File: Packs/EWS/Integrations/EWSO365/EWSO365.py (hunk 11 above)
Comment (JasBeilin) [reply Packs/EWS/Integrations/EWSO365/EWSO365.py:2171]: yes
image

**Review by yuvalbenshalom — APPROVED** [review]

File: Packs/EWS/Integrations/EWSO365/EWSO365.py (hunk 11 above)
Comment (guykeller) [reply Packs/EWS/Integrations/EWSO365/EWSO365.py:2171]:
```suggestion
        most_recent_incident_time = incident.get("occurred", last_fetch_time)
```

File: Packs/EWS/Integrations/EWSO365/EWSO365.py [hunk 13]
```diff
@@ -2143,35 +2142,51 @@ def fetch_emails_as_incidents(client: EWSClient, last_run):
...
+            last_incident_run_time = last_incident_run_time.ewsformat()
+
+        if isinstance(last_fetch_time, EWSDateTime):
+            last_fetch_time = last_fetch_time.ewsformat()
+
+        demisto.debug(
+            f'#### last_incident_time: {last_incident_run_time}({type(last_incident_run_time)}).'
+            f'last_fetch_time: {last_fetch_time}({type(last_fetch_time)}) ####')
+
+        if not last_incident_run_time or not last_fetch_time or last_incident_run_time > last_fetch_time:
```
Comment (guykeller) [inline Packs/EWS/Integrations/EWSO365/EWSO365.py:2183]: Please add an explaining comment here for future contributors

File: Packs/EWS/Integrations/EWSO365/EWSO365.py [hunk 14]
```diff
@@ -2143,35 +2142,51 @@ def fetch_emails_as_incidents(client: EWSClient, last_run):
...
         incidents = []
         incident: Dict[str, str] = {}
+        demisto.debug(f'{APP_NAME} - Started fetch with {len(last_emails)} at {last_run.get(LAST_RUN_TIME)}')
+        current_fetch_ids = set()
         for item in last_emails:
             if item.message_id:
-                ids.append(item.message_id)
+                current_fetch_ids.add(item.message_id)
                 incident = parse_incident_from_item(item)
                 incidents.append(incident)
```
[the remark this answers is not in the document]
Comment (JasBeilin) [reply Packs/EWS/Integrations/EWSO365/EWSO365.py:2162]: No, since this is only taking the already distinct id's and parse the message to incident format.

**Pull request merged**